xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -0,0 +1,315 @@
+type xdsClientInterface interface {
+	ReportLoad(server string) (*load.Store, func())
+	Close()
+}
+
+type clusterImplBalancer struct {
+	cc     balancer.ClientConn
+	closed *grpcsync.Event
+
+	logger *grpclog.PrefixLogger
+	xdsC   xdsClientInterface
+
+	config  *lbConfig
+	childLB balancer.Balancer
+
+	cancelLoadReport func()
+	clusterName      string
+	edsServiceName   string
+	lrsServerName    string
+	loadWrapper      *utils.LoadStoreWrapper
+
+	// childState/drops/requestCounter can only be accessed in run(). And run()
+	// is the only goroutine that sends picker to the parent ClientConn. All
+	// requests to update picker need to be sent to pickerUpdateCh.
+	childState      balancer.State
+	drops           []*dropper
+	requestCounter  *xdsclient.ServiceRequestsCounter
+	requestCountMax uint32
+	pickerUpdateCh  *buffer.Unbounded
+}
+
+// updateLoadStore checks the config for load store, and decides whether it
+// needs to restart the load reporting stream.
+func (cib *clusterImplBalancer) updateLoadStore(newConfig *lbConfig) error {
+	var (
+		restartLoadReport           bool
+		updateLoadClusterAndService bool
+	)
+
+	// ClusterName is different, restart. ClusterName is from ClusterName and
+	// EdsServiceName.
+	if cib.clusterName != newConfig.Cluster {
+		updateLoadClusterAndService = true
+		cib.clusterName = newConfig.Cluster
+	}
+	if cib.edsServiceName != newConfig.EDSServiceName {
+		updateLoadClusterAndService = true
+		cib.edsServiceName = newConfig.EDSServiceName
+	}
+	if updateLoadClusterAndService {
+		// This updates the clusterName and serviceName that will reported for the
+		// loads. The update here is too early, the perfect timing is when the
+		// picker is updated with the new connection. But from this balancer's point
+		// of view, it's impossible to tell.
+		//
+		// On the other hand, this will almost never happen. Each LRS policy
+		// shouldn't get updated config. The parent should do a graceful switch when
+		// the clusterName or serviceName is changed.
+		cib.loadWrapper.UpdateClusterAndService(cib.clusterName, cib.edsServiceName)
+	}
+
+	// Check if it's necessary to restart load report.
+	var newLRSServerName string
+	if newConfig.LRSLoadReportingServerName != nil {
+		newLRSServerName = *newConfig.LRSLoadReportingServerName
+	}
+	if cib.lrsServerName != newLRSServerName {
+		// LrsLoadReportingServerName is different, load should be report to a
+		// different server, restart.
+		restartLoadReport = true
+		cib.lrsServerName = newLRSServerName
+	}
+	if restartLoadReport {
+		if cib.cancelLoadReport != nil {
+			cib.cancelLoadReport()
+			cib.cancelLoadReport = nil
+		}
+		var loadStore *load.Store
+		if cib.xdsC != nil {
+			loadStore, cib.cancelLoadReport = cib.xdsC.ReportLoad(cib.lrsServerName)
+		}
+		cib.loadWrapper.UpdateLoadStore(loadStore)
+	}
+
+	return nil
+}
+
+func (cib *clusterImplBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
```

Comment:
Do these methods which implement the `balancer.Balancer` interface need to check `closed.HasFired()` at their top?